cibuild.sh

```diff
@@ -0,0 +1,171 @@
+           fi; \
+           p32-gcc --version; \
+           ")
+path+=("$prebuilt/pinguino-compilers/linux64/p32/bin")
+
+# risc-v gcc toolchain
+install+=("if [ ! -d \"$prebuilt/riscv64-unknown-elf-gcc/bin\" ]; then \
+             cd $prebuilt; \
+             wget https://static.dev.sifive.com/dev-tools/riscv64-unknown-elf-gcc-8.3.0-2019.08.0-x86_64-linux-ubuntu14.tar.gz; \
+             tar zxf riscv64-unknown-elf-gcc-8.3.0-2019.08.0-x86_64-linux-ubuntu14.tar.gz; \
+             mv riscv64-unknown-elf-gcc-8.3.0-2019.08.0-x86_64-linux-ubuntu14 riscv64-unknown-elf-gcc; \
+             rm riscv64-unknown-elf-gcc-8.3.0-2019.08.0-x86_64-linux-ubuntu14.tar.gz; \
+           fi; \
+           riscv64-unknown-elf-gcc --version; \
+           ")
+path+=("$prebuilt/riscv64-unknown-elf-gcc/bin")
+
+# ccache
+install+=("if [ ! -d \"$prebuilt/ccache/bin\" ]; then \
+             cd $prebuilt; \
+             wget https://github.com/ccache/ccache/releases/download/v3.7.7/ccache-3.7.7.tar.gz; \
+             tar zxf ccache-3.7.7.tar.gz; \
+             cd ccache-3.7.7; ./configure --prefix=$prebuilt/ccache; make; make install; \
+             cd $prebuilt; rm -rf ccache-3.7.7; rm ccache-3.7.7.tar.gz; \
+             ln -sf $prebuilt/ccache/bin/ccache $prebuilt/ccache/bin/gcc; \
+             ln -sf $prebuilt/ccache/bin/ccache $prebuilt/ccache/bin/g++; \
+             ln -sf $prebuilt/ccache/bin/ccache $prebuilt/ccache/bin/arm-none-eabi-gcc; \
+             ln -sf $prebuilt/ccache/bin/ccache $prebuilt/ccache/bin/arm-none-eabi-g++; \
+             ln -sf $prebuilt/ccache/bin/ccache $prebuilt/ccache/bin/p32-gcc; \
+             ln -sf $prebuilt/ccache/bin/ccache $prebuilt/ccache/bin/riscv64-unknown-elf-gcc; \
+             ln -sf $prebuilt/ccache/bin/ccache $prebuilt/ccache/bin/riscv64-unknown-elf-g++; \
+           fi; \
+           ccache --version; \
+           export USE_CCACHE=1; \
+           export CCACHE_DIR=$prebuilt/ccache/.ccache; \
+           ccache -M 5G; \
+           ")
+path+=("$prebuilt/ccache/bin")
+
+function usage {
+  echo ""
+  echo "USAGE: $0 check | full"
+  echo "       $0 -h"
+  echo ""
+  echo "Where:"
+  echo "  check:   do check build CI"
+  echo "  full:    do full build CI"
+  echo "  -h will show this help test and terminate"
+  echo ""
+  exit 1
+}
+
+function setup_repos {
+  if [ -d "$nuttx" ]; then
+    cd $nuttx; git pull
+  else
+    git clone https://github.com/apache/incubator-nuttx.git $nuttx
+  fi
+
+  if [ -d "$apps" ]; then
+    cd $apps; git pull
+  else
+    git clone https://github.com/apache/incubator-nuttx-apps.git $apps
+  fi
+}
+
+function install_tools {
+  for ((i=0; i<${#install[@]}; ++i)); do
+    export PATH=${path[$i]}:$PATH
+    eval ${install[$i]}
+  done
+}
+
+function run_builds {
+  local JOBS=`grep -c ^processor /proc/cpuinfo`
+
+  $nuttx/tools/testbuild.sh -si -j $JOBS $WD/testlist/${build}list.dat
+  if [ $? != 0 ]; then
+    echo "ERROR: $BUILD build failed."
```

Comment:
Add the error code to message: `echo "ERROR: $BUILD build failed (error $?)"`

Reply:
Return $? would be ok, now testbuild.sh would just return fail value 0 or 1 in fact. Since testbuild.sh keep on going build even build fail in some build stage, error values would be overwrited. So it just keep the fail value as failed 1 or success 0.

Reply:
But the final error value would not be overwritten, it isn't? So why not add it in the message?

Reply:
Right, thanks, I should also make a PR to update testbuild.sh to correct fail value as the final errno value.

Reply:
done, and testbuild.sh PR also upstream merged